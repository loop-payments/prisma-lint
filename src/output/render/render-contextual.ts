import chalk from 'chalk';

import type { Violation } from '#src/common/violation.js';
import { keyViolationListPairs } from '#src/output/render/render-util.js';

export const renderViolationsContextual = (
  sourceCode: string,
  violations: Violation[],
): string[] => {
  const pairs = keyViolationListPairs(violations);
  const lines = pairs.flatMap(([key, violations]) => {
    const first = violations[0];
    const { fileName } = first;
    const rawLocation =
      first.field?.location ?? first.model?.location ?? first.enum?.location;
    if (!rawLocation) {
      throw new Error('No location');
    }
    const {
      startLine,
      startColumn,
      endLine,
      endColumn,
      startOffset,
      endOffset,
    } = rawLocation;
    if (
      !startLine ||
      !startColumn ||
      !endLine ||
      !endColumn ||
      !startOffset ||
      !endOffset
    ) {
      throw new Error('No line or column');
    }
    const lines = sourceCode.split('\n');
    const containingLine = lines[startLine - 1];
    const pointer =
      ' '.repeat(startColumn - 1) + '^'.repeat(endColumn - startColumn + 1);
    return [
      '',
      `${fileName}:${startLine}:${startColumn} ${chalk.gray(`${key}`)}`,
      `${containingLine}`,
      `${pointer}`,
    ].concat(
      violations.flatMap((violation) => {
        const { ruleName, message } = violation;
        return [
          `  ${chalk.red('error')} ${message} ${chalk.gray(`${ruleName}`)}`,
        ];
      }),
    );
  });
  return lines;
};
