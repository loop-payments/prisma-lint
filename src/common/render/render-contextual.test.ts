import { renderViolationsContextual } from '#src/common/render/render-contextual.js';
import {
  MOCK_SOURCE_CODE,
  MOCK_VIOLATIONS,
} from '#src/common/render/render-test-util.js';

describe('render contextual', () => {
  it('matches snapshot', () => {
    const result = renderViolationsContextual(
      MOCK_SOURCE_CODE,
      MOCK_VIOLATIONS,
    );
    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });
});
