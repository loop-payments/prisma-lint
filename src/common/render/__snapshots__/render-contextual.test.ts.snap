// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render contextual matches snapshot 1`] = `
[
  "",
  "test.prisma:1:1 [90mFoo[39m",
  "model Foo {",
  "^^^^^^^^^",
  "  [31merror[39m Fake model rule violation message [90mfake-model-rule-name[39m",
  "",
  "test.prisma:2:3 [90mFoo.displayName[39m",
  "  id     Int    @id",
  "  ^^^^^^^^^^",
  "  [31merror[39m Fake field rule violation message [90mfake-field-rule-name[39m",
]
`;
